src/include/storage/index/bwtree_index.h

```diff
@@ -31,61 +28,178 @@ class BwTreeIndex final : public Index {
     KeyType index_key;
     index_key.SetFromProjectedRow(key, metadata_);
-    bwtree_->GetValue(index_key, *value_list);
+
+    // Perform lookup in BwTree
+    bwtree_->GetValue(index_key, results);
+
+    // Avoid resizing our value_list, even if it means over-provisioning
+    value_list->reserve(results.size());
+
+    // Perform visibility check on result
+    for (const auto &result : results) {
+      if (IsVisible(txn, result)) value_list->emplace_back(result);
+    }
+
+    TERRIER_ASSERT(GetConstraintType() == ConstraintType::DEFAULT ||
+                       (GetConstraintType() == ConstraintType::UNIQUE && value_list->size() <= 1),
+                   "Invalid number of results for unique index.");
   }
 
-  void Scan(const ProjectedRow &low_key, const ProjectedRow &high_key, std::vector<TupleSlot> *value_list) final {
-    TERRIER_ASSERT(
-        value_list->empty(),
-        "Result set should begin empty. This can be changed in the future if index scan behavior requires it.");
+  void ScanAscending(const transaction::TransactionContext &txn, const ProjectedRow &low_key,
+                     const ProjectedRow &high_key, std::vector<TupleSlot> *value_list) final {
+    TERRIER_ASSERT(value_list->empty(), "Result set should begin empty.");
+
+    // Build search keys
     KeyType index_low_key, index_high_key;
     index_low_key.SetFromProjectedRow(low_key, metadata_);
     index_high_key.SetFromProjectedRow(high_key, metadata_);
 
-    for (auto scan_itr = bwtree_->Begin(index_low_key);
-         !scan_itr.IsEnd() && (bwtree_->KeyCmpLessEqual(scan_itr->first, index_high_key)); scan_itr++) {
-      value_list->emplace_back(scan_itr->second);
+    // Perform lookup in BwTree
+    auto scan_itr = bwtree_->Begin(index_low_key);
+    while (!scan_itr.IsEnd() && (bwtree_->KeyCmpLessEqual(scan_itr->first, index_high_key))) {
+      // Perform visibility check on result
+      if (IsVisible(txn, scan_itr->second)) value_list->emplace_back(scan_itr->second);
+      scan_itr++;
+    }
+  }
+
+  void ScanDescending(const transaction::TransactionContext &txn, const ProjectedRow &low_key,
+                      const ProjectedRow &high_key, std::vector<TupleSlot> *value_list) final {
+    TERRIER_ASSERT(value_list->empty(), "Result set should begin empty.");
+
+    // Build search keys
+    KeyType index_low_key, index_high_key;
+    index_low_key.SetFromProjectedRow(low_key, metadata_);
+    index_high_key.SetFromProjectedRow(high_key, metadata_);
+
+    // Perform lookup in BwTree
+    auto scan_itr = bwtree_->Begin(index_high_key);
+    // Back up one element if we didn't match the high key
+    if (scan_itr.IsEnd() || bwtree_->KeyCmpGreater(scan_itr->first, index_high_key)) scan_itr--;
+
+    while (!scan_itr.IsREnd() && (bwtree_->KeyCmpGreaterEqual(scan_itr->first, index_low_key))) {
+      // Perform visibility check on result
+      if (IsVisible(txn, scan_itr->second)) value_list->emplace_back(scan_itr->second);
+      scan_itr--;
```

Comment:
Have you checked whether reverse iteration is faster than just wrapping `ScanAscending` with an explicit reversing of the vector?  I seem to recall discussion that this particular call is expensive.